src/neo/Network/P2P/Payloads/Transaction.cs

```diff
@@ -295,17 +295,36 @@ public virtual VerifyResult VerifyStateDependent(StoreView snapshot, Transaction
                 if (!attribute.Verify(snapshot, this))
                     return VerifyResult.Invalid;
             long net_fee = NetworkFee - Size * NativeContract.Policy.GetFeePerByte(snapshot);
-            if (!this.VerifyWitnesses(snapshot, net_fee, WitnessFlag.StateDependent))
-                return VerifyResult.Invalid;
+
+            UInt160[] hashes = GetScriptHashesForVerifying(snapshot);
+            if (hashes.Length != witnesses.Length) return VerifyResult.Invalid;
+            for (int i = 0; i < hashes.Length; i++)
+            {
+                if (witnesses[i].VerificationScript.IsSignatureContract())
+                    net_fee -= (ApplicationEngine.OpCodePrices[OpCode.PUSHDATA1] * 2 + ApplicationEngine.OpCodePrices[OpCode.PUSHNULL] + ApplicationEngine.OpCodePrices[OpCode.SYSCALL] + ApplicationEngine.ECDsaVerifyPrice);
+                else if (witnesses[i].VerificationScript.IsMultiSigContract(out int m, out int n))
+                    net_fee -= (ApplicationEngine.OpCodePrices[OpCode.PUSHDATA1] * (m + n) + ApplicationEngine.OpCodePrices[OpCode.PUSHINT8] * 2 + ApplicationEngine.OpCodePrices[OpCode.PUSHNULL] + ApplicationEngine.OpCodePrices[OpCode.SYSCALL] + ApplicationEngine.ECDsaVerifyPrice * n);
+                else
+                {
+                    if (!this.VerifyWitness(null, hashes[i], witnesses[i], net_fee, out long fee))
+                        return VerifyResult.Invalid;
+                    net_fee -= fee;
+                }
+                if (net_fee < 0) return VerifyResult.InsufficientFunds;
+            }
             return VerifyResult.Succeed;
         }
 
         public virtual VerifyResult VerifyStateIndependent()
         {
             if (Size > MaxTransactionSize)
                 return VerifyResult.Invalid;
-            if (!this.VerifyWitnesses(null, NetworkFee, WitnessFlag.StateIndependent))
-                return VerifyResult.Invalid;
+            UInt160[] hashes = GetScriptHashesForVerifying(null);
+            if (hashes.Length != witnesses.Length) return VerifyResult.Invalid;
+            for (int i = 0; i < hashes.Length; i++)
+                if (witnesses[i].VerificationScript.IsStandardContract())
+                    if (!this.VerifyWitness(null, hashes[i], witnesses[i], SmartContract.Helper.MaxVerificationGas, out _))
```

Comment:
If it is a standard contract, we can verify the signature(s) directly. We don't need to launch the vm.

Reply:
Sending `null` as the snapshot is weird.

Reply:
> If it is a standard contract, we can verify the signature(s) directly. We don't need to launch the vm.

No difference in performance, we have tested before.

> Sending null as the snapshot is weird.

But if we pass a snapshot, it's not state independent.

Reply:
Or we can pass the snapshot, and add some comments here.

Reply:
It seems we can't pass the snapshot in `TransactionRouter`.

https://github.com/neo-project/neo/blob/6b139e31ef658f473692bf3f16e3835d733b59ac/src/neo/Ledger/TransactionRouter.cs#L25

And for standard witnesses, they should not use snapshot.